src/nemos/solvers/_abstract_solver.py

```diff
@@ -0,0 +1,109 @@
+"""Base class defining the interface for solvers that can be used by `BaseRegressor`."""
+
+import abc
+from typing import Any, Callable, Generic, NamedTuple
+
+from ..regularizer import Regularizer
+from ..typing import Params, SolverState, StepResult
+
+# TODO: If we want to accept solvers that implement this interface,
+# but are not implemented as a subclass of AbstractSolver,
+# we could just create a protocol listing the methods.
+
+
+class OptimizationInfo(NamedTuple):
+    """Basic diagnostic information about finished optimization runs."""
+
+    # Not all JAXopt solvers store the function value.
+    # Not sure how confusing it is to store None when it's not present
+    # because it can be confused with a diverged optimization.
+    function_val: float | None
```

Comment:
what happens when there's a diverged optimization? that feels like it should maybe be a NaN, rather than None.

Reply:
I think diverged optimizations give NaN which should be covered by float.
None is added because some solvers don't provide it and we don't evaluate it because it is potentially costly, so in that case it's missing.
So NaN is "we have this value, it's just bad", None is "we don't have this value".

I also think it's confusing (hence the comment above it). Do you have a suggestion?